vault/resource_plugin.go

```diff
@@ -0,0 +1,251 @@
+			fieldSHA256: {
+				Type:        schema.TypeString,
+				Description: "SHA256 sum of the plugin binary.",
+				Required:    true,
+			},
+			fieldCommand: {
+				Type:        schema.TypeString,
+				Description: "Command to execute the plugin, relative to the plugin_directory.",
+				Required:    true,
+			},
+			fieldArgs: {
+				Type:        schema.TypeList,
+				Description: "List of additional arguments to pass to the plugin.",
+				Optional:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			fieldEnv: {
+				Type:        schema.TypeList,
+				Description: "List of additional environment variables to run the plugin with in KEY=VALUE form.",
+				Optional:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Sensitive: true,
+			},
+			fieldOCIImage: {
+				Type:        schema.TypeString,
+				Description: "OCI image to run. If specified, setting command, args, and env will update the container's entrypoint, args, and environment variables (append-only) respectively.",
+				Optional:    true,
+			},
+			fieldRuntime: {
+				Type:        schema.TypeString,
+				Description: "Vault plugin runtime to use if oci_image is specified.",
+				Optional:    true,
+			},
+		},
+	}
+}
+
+func pluginWrite(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, err := provider.GetClient(d, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	pluginType, err := api.ParsePluginType(d.Get(consts.FieldType).(string))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	name := d.Get(consts.FieldName).(string)
+	version := d.Get(consts.FieldVersion).(string)
+	id := fmt.Sprintf("%s/%s", pluginType, name)
+	if version != "" {
+		id = fmt.Sprintf("%s/%s", id, version)
+	}
+
+	if diagErr := versionedPluginsSupported(meta, version); diagErr != nil {
+		return diagErr
+	}
+
+	ociImage := d.Get(fieldOCIImage).(string)
+	runtime := d.Get(fieldRuntime).(string)
+	if diagErr := containerizedPluginsSupported(meta, ociImage, runtime); diagErr != nil {
+		return diagErr
+	}
+
+	log.Printf("[DEBUG] Writing plugin %q", id)
+	err = client.Sys().RegisterPluginWithContext(ctx, &api.RegisterPluginInput{
+		Type:     pluginType,
+		Name:     name,
+		Version:  version,
+		SHA256:   d.Get(fieldSHA256).(string),
+		Command:  d.Get(fieldCommand).(string),
+		Args:     util.ToStringArray(d.Get(fieldArgs).([]interface{})),
+		Env:      util.ToStringArray(d.Get(fieldEnv).([]interface{})),
+		OCIImage: ociImage,
+		Runtime:  runtime,
+	})
+	if err != nil {
+		return diag.Errorf("error updating plugin %q: %s", id, err)
+	}
+	log.Printf("[DEBUG] Wrote plugin %q", id)
+
+	d.SetId(id)
+
+	return pluginRead(ctx, d, meta)
+}
+
+func pluginRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return diag.FromErr(e)
+	}
+
+	var typ, name, version string
+	parts := strings.Split(d.Id(), "/")
```

Comment:
What happens here if the name has a `/` character in it? For example, `secret/prod/gcp/v0.17.0`. Here, while the type will be correctly set, I believe the logic here  would set the name as `prod` and version as `gcp/v0.17.0`. Should we instead use some form of a regex here?

Reply:
In the past, I have employed length checks for IDs and paths similar to the above, but learned the hard way that we can never predict what the user might use as a name for a secret or in this case, a plugin. If there are some validations the Vault code already performs in order to prevent `/`s in names then this is not a concern. Otherwise, maybe a regex [like this](https://github.com/hashicorp/terraform-provider-vault/blob/main/vault/resource_ad_secret_library.go#L19-L20) might be more robust?

Reply:
Oof, I forgot plugin names can have slashes in them, good catch. It would actually error out because `len(parts) > 3` but we definitely should handle that better.

Reply:
I see what you mean. I think the main objective for the ID is to be unique and parseable on our end, so `<version>/<type>/<name>` does at least achieve that. Wondering if a prefix of something like `sys/plugins/<version>/<type>/<name>` makes it a bit more readable? Starting an ID with a semantic version string shouldn't be a problem, but can seem out of place since we have no precedents for that in the repo.

Reply:
I've reworked this now to `:type/version/:version/name/:name` to match the pattern that secret sync associations used. Thanks for the feedback!